pyteal/ast/opup.py

```diff
@@ -87,41 +106,59 @@ def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
 
         self.mode = mode
 
-    def _construct_itxn(self) -> Expr:
+    def _construct_itxn(self, inner_fee: Expr = None) -> Expr:
+        fields: dict[TxnField, Expr | list[Expr]] = {
+            TxnField.type_enum: TxnType.ApplicationCall
+        }
+
+        # If an inner_fee is specified
+        # add it to the transaction fields
+        if inner_fee is not None:
+            require_type(inner_fee, TealType.uint64)
+            fields[TxnField.fee] = inner_fee
+
         if self.mode == OpUpMode.Explicit:
-            return Seq(
-                InnerTxnBuilder.Begin(),
-                InnerTxnBuilder.SetFields(
-                    {
-                        TxnField.type_enum: TxnType.ApplicationCall,
-                        TxnField.application_id: self.target_app_id,
-                    }
-                ),
-                InnerTxnBuilder.Submit(),
-            )
+            fields |= {TxnField.application_id: self.target_app_id}
         else:
-            return Seq(
-                InnerTxnBuilder.Begin(),
-                InnerTxnBuilder.SetFields(
-                    {
-                        TxnField.type_enum: TxnType.ApplicationCall,
-                        TxnField.on_completion: OnComplete.DeleteApplication,
-                        TxnField.approval_program: ON_CALL_APP,
-                        TxnField.clear_state_program: ON_CALL_APP,
-                    }
-                ),
-                InnerTxnBuilder.Submit(),
-            )
+            fields |= {
+                TxnField.on_completion: OnComplete.DeleteApplication,
+                TxnField.approval_program: ON_CALL_APP,
+                TxnField.clear_state_program: ON_CALL_APP,
+            }
+
+        return InnerTxnBuilder.Execute(fields)
 
-    def ensure_budget(self, required_budget: Expr) -> Expr:
+    def ensure_budget(
+        self, required_budget: Expr, fee_source: OpUpFeeSource = OpUpFeeSource.Any
+    ) -> Expr:
         """Ensure that the budget will be at least the required_budget.
 
+        Args:
+            required_budget: minimum op-code budget to ensure for the
+                upcoming execution.
+            inner_fee (optional): controls the behavior of setting the fee for
```

Comment:
@barnjamin Friendly visual reminder (also for me) to update docs once we finalize naming conventions.

Reply:
still need to get back to this but marking as ready to review to get any more feedback on code